File: typescript/packages/core/src/core/gslides/read.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import type { GSlidesAccessor } from '../../accessor/gslides.ts'
import type { IndexCacheStore } from '../../cache/index/store.ts'
import type { PathSpec } from '../../types.ts'
import { enoent } from '../../utils/errors.ts'
import { slidesBase, type TokenManager, googleGet } from '../google/client.ts'
import { resolveEntry } from '../hierarchy/probe.ts'
import { makeRead } from '../hierarchy/read.ts'
import type { ScopeMatch } from '../hierarchy/scope.ts'
import { compactJsonBytes } from '../render/json.ts'
import { readdir } from './readdir.ts'
import { detectScope } from './scope.ts'

export async function readPresentation(
  tm: TokenManager,
  presentationId: string,
): Promise<Uint8Array> {
  const url = `${slidesBase(tm)}/presentations/${presentationId}`
  const data = await googleGet(tm, url)
  return compactJsonBytes(data)
}

async function readFile(
  accessor: GSlidesAccessor,
  _match: ScopeMatch,
  path: PathSpec,
  index?: IndexCacheStore,
): Promise<Uint8Array> {
  const entry = await resolveEntry(readdir, accessor, path, index)
  if (entry === null) throw enoent(path.virtual)
  return readPresentation(accessor.tokenManager, entry.id)
}

export const read = makeRead<GSlidesAccessor>(detectScope, { file: readFile })

export async function* stream(
  accessor: GSlidesAccessor,
  path: PathSpec,
  index?: IndexCacheStore,
): AsyncIterable<Uint8Array> {
  yield await read(accessor, path, index)
}

File: python/mirage/core/gslides/read.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

from mirage.accessor.gslides import GSlidesAccessor
from mirage.cache.index import IndexCacheStore
from mirage.core.gslides.client import TokenManager, google_get, slides_base
from mirage.core.gslides.readdir import readdir
from mirage.core.gslides.scope import detect_scope
from mirage.core.hierarchy.probe import resolve_entry
from mirage.core.hierarchy.read import make_read
from mirage.core.hierarchy.scope import ScopeMatch
from mirage.core.render.json import compact_json_bytes
from mirage.types import PathSpec
from mirage.utils.errors import enoent


async def read_presentation(token_manager: TokenManager,
                            presentation_id: str) -> bytes:
    url = f"{slides_base(token_manager)}/presentations/{presentation_id}"
    data = await google_get(token_manager, url)
    return compact_json_bytes(data)


async def _read_file(accessor: GSlidesAccessor, match: ScopeMatch,
                     path: PathSpec, index: IndexCacheStore) -> bytes:
    entry = await resolve_entry(readdir, accessor, path, index)
    if entry is None:
        raise enoent(path.virtual)
    return await read_presentation(accessor.token_manager, entry.id)


read = make_read(detect_scope, readers={"file": _read_file})
